### PR TITLE
feat : Implemented new feature for devo-run-query, devo-get-alerts an…

### PR DESCRIPTION
**Pull request by manas-metron**
…d devo-multi-table-query.

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Devo/Integrations/Devo_v2/Devo_v2.py [hunk 1]
```diff
@@ -520,6 +520,41 @@ def fetch_incidents():
...
+        - results (list): A list of dictionaries representing rows of data.
+        - fields_to_view (list): A field names to include.
+
+    Returns:
+        list: A new list of dictionaries with only the specified fields.
+
+    Raises:
+        ValueError: If the fields_to_view contains invalid column names.
+    """
+    if not fields_to_view or not results:
```
Comment (sahil-metron) [inline Packs/Devo/Integrations/Devo_v2/Devo_v2.py]: @manas-metron what happens if both of these parameters are empty ?

**Review by sahil-metron — CHANGES_REQUESTED** [review]
Please check the comment.

File: Packs/Devo/Integrations/Devo_v2/Devo_v2.py (hunk 1 above)
Comment (manas-metron) [reply Packs/Devo/Integrations/Devo_v2/Devo_v2.py]: "results" is not a parameter. "results" is the actual response from the devo.
If "fields_to_view" is not passed, then we just return the actual response.

**Review by sahil-metron — APPROVED** [review]

**Review by sahil-metron — APPROVED** [review]

**Review by kapil-metron — APPROVED** [review]

**Review by sahil-metron — APPROVED** [review]

File: Packs/Devo/Integrations/Devo_v2/Devo_v2.py [hunk 2]
```diff
@@ -520,6 +520,46 @@ def fetch_incidents():
...
+    first_dict = results[0]
+    missing_columns = set(fields_to_view_list) - set(first_dict)
+    if missing_columns:
+        raise ValueError(f"Fields {list(missing_columns)} not found in query result")
+
+    filtered_results = []
+
+    for result in results:
+        filtered_result = {
+            column: tuple(result.get(column)) if isinstance(result.get(column), list) else result.get(column)
```
Comment (pavel-kalmykov) [inline Packs/Devo/Integrations/Devo_v2/Devo_v2.py]: **question:** Could you please explain why we are transforming lists into tuples here?

File: Packs/Devo/Integrations/Devo_v2/Devo_v2_test.py [hunk 3]
```diff
@@ -319,7 +336,28 @@ def test_get_alerts(mock_query_results, mock_args_results):
...
-    assert results[0]["Contents"][0]["engine"] == "CPU_Usage_Alert"
+    assert results[0]["Contents"][0]["context"] == "CPU_Usage_Alert"
+    # Check if all expected columns are present in the dictionary
+    # Convert fields_to_view from a list to a comma-separated string
+    expected_columns = ','.join(field.strip() for field in MOCK_ALERT_ARGS['fields_to_view'].split(','))
+    result = results[0]["Contents"][0]
+    assert all(column in result for column in expected_columns.split(',')), (
+        f"Not all columns present in the dictionary. Missing columns: "
+        f"{', '.join(column for column in expected_columns.split(',') if column not in result)}"
+    )
```
Comment (pavel-kalmykov) [inline Packs/Devo/Integrations/Devo_v2/Devo_v2_test.py]: **suggestion(blocking):** Do not add extra assertions to an already-working test, as each test case must test only one thing (they must be atomic). Instead, create new cases.

Note that the function name for this test case is not much explanatory. A good test name must contain the SUT, entered input and expected output. You might want to revise this in other test cases as well.

File: Packs/Devo/Integrations/Devo_v2/Devo_v2_test.py [hunk 4]
```diff
@@ -319,7 +336,28 @@ def test_get_alerts(mock_query_results, mock_args_results):
...
+@patch("Devo_v2.READER_OAUTH_TOKEN", MOCK_READER_OAUTH_TOKEN, create=True)
+@patch("Devo_v2.demisto.args")
+@patch("Devo_v2.ds.Reader.query")
+def test_get_alerts_with_empty_fields_to_view_param(mock_query_results, mock_args_results):
+    mock_query_results.return_value = copy.deepcopy(MOCK_QUERY_RESULTS)
+    mock_args_results.return_value = MOCK_ALERT_ARGS_EMPTY_FIELDS_TO_VIEW_PRAM
+    try:
+        get_alerts_command(OFFSET, ITEMS_PER_PAGE)
+    except ValueError as e:
+        assert "fields_to_view cannot be an empty string!" in str(e), f"Unexpected exception message: {e}"
```
Comment (pavel-kalmykov) [inline Packs/Devo/Integrations/Devo_v2/Devo_v2_test.py]: **suggestion(blocking):** As you are using pytest, you can use `pytest.raises` to check when a callable throws an exception. Read more [here](https://docs.pytest.org/en/7.1.x/how-to/assert.html).

File: Packs/Devo/Integrations/Devo_v2/Devo_v2.py [hunk 5]
```diff
@@ -598,6 +644,9 @@ def get_alerts_command(offset, items):
     linq_base = demisto.args().get("linqLinkBase", None)
     user_alert_table = demisto.args().get("table_name", None)
     user_prefix = demisto.args().get("prefix", "")
+    fields_to_view = demisto.args().get("fields_to_view", None)
+    if fields_to_view == "":
+        raise ValueError("fields_to_view cannot be an empty string!")
```
Comment (pavel-kalmykov) [inline Packs/Devo/Integrations/Devo_v2/Devo_v2.py]: **comment(non-blocking):** This snippet of code and the `if fields_to_view: results [...]` is repeated three times, each per command. When requesting this new functionality, I asked you to consider refactoring these commands to simplify their logic. Please, reconsider it for the future.

File: Packs/Devo/Integrations/Devo_v2/Devo_v2_test.py
Comment (pavel-kalmykov) [inline Packs/Devo/Integrations/Devo_v2/Devo_v2_test.py:1]: **suggestion(blocking):** Here are some additional test cases I asked you to design:

1. When receiving a field that is not in the result, you must check that the error is something like `"field 'abcd' not found in query result"`.
2. When a field is repeated, additional entries will be ignored. In other words, the column will be displayed only once.

You don't have to implement these cases for each command!

File: Packs/Devo/Integrations/Devo_v2/Devo_v2.yml [hunk 6]
```diff
@@ -115,6 +115,8 @@ script:
       defaultValue: "true"
     - name: linqLinkBase
       description: Overrides the global Devo base domain for linq linking.
+    - name: fields_to_view
+      description: A field names to include in result.
```
Comment (pavel-kalmykov) [inline Packs/Devo/Integrations/Devo_v2/Devo_v2.yml]: **issue:** I'd like you to change the description to:
```suggestion
      description: The subset of fields (separated by a comma) that you want to display from the query result. Use this if you want to filter out unwanted columns in your result.
```
Do it for all the commands.

**Review by pavel-kalmykov — CHANGES_REQUESTED** [review]
Here are some quick things I was able to notice. Please, review them carefully and let me know if you have any further questions.

**Review by pavel-kalmykov — CHANGES_REQUESTED** [review]
Here are some quick things I was able to notice. Please, review them carefully and let me know if you have any further questions.

File: Packs/Devo/Integrations/Devo_v2/Devo_v2.py [hunk 7]
```diff
@@ -534,6 +534,9 @@ def filter_results_by_fields(results, fields_to_view_string):
     Raises:
         ValueError: If the fields_to_view_string contains invalid column names.
     """
+    if fields_to_view_string == "":
+        raise ValueError("fields_to_view cannot be an empty.")
```
Comment (pavel-kalmykov) [inline Packs/Devo/Integrations/Devo_v2/Devo_v2.py]: **suggestion:**
```suggestion
        raise ValueError("fields_to_view cannot be empty.")
```
Additionally, I would `strip()` the `fields_to_view_string`.

File: Packs/Devo/Integrations/Devo_v2/Devo_v2_test.py [hunk 8]
```diff
@@ -371,25 +412,17 @@ def test_run_query(mock_query_results, mock_args_results):
...
-def test_run_query_with_empty_fields_to_view_param(mock_query_results, mock_args_results):
+def test_run_query_with_invalid_column_name(mock_query_results, mock_args_results):
     mock_query_results.return_value = copy.deepcopy(MOCK_QUERY_RESULTS)
-    mock_args_results.return_value = MOCK_QUERY_ARGS_EMPTY_FIELDS_TO_VIEW_PRAM
-    try:
+    mock_args_results.return_value = MOCK_QUERY_ARGS_INVALIDE_COLUMN_NAME
+    with pytest.raises(ValueError, match=re.escape("Fields ['abcd'] not found in query result")):
         run_query_command(OFFSET, ITEMS_PER_PAGE)
-    except ValueError as e:
-        assert "fields_to_view cannot be an empty string!" in str(e), f"Unexpected exception message: {e}"
```
Comment (pavel-kalmykov-razgovorov) [inline Packs/Devo/Integrations/Devo_v2/Devo_v2_test.py]: **question(blocking):** Does this test run? I don't see this error thrown in the code anywhere.

**Review by pavel-kalmykov — CHANGES_REQUESTED** [review]
Here are two little nitpicks I found this time:

File: Packs/Devo/Integrations/Devo_v2/Devo_v2_test.py (hunk 8 above)
Comment (pavel-kalmykov-razgovorov) [reply Packs/Devo/Integrations/Devo_v2/Devo_v2_test.py]: My bad; I just found it.

File: Packs/Devo/Integrations/Devo_v2/Devo_v2.py (hunk 7 above)
Comment (manas-metron) [reply Packs/Devo/Integrations/Devo_v2/Devo_v2.py]: If we strip the empty fields_to_view_string, we won't be able to raise the error.I already checked that.

**Review by pavel-kalmykov — APPROVED** [review]
LGTM!

**Pull request closed without merging**